diagnostics.go

```diff
@@ -0,0 +1,323 @@
+// Copyright 2017 Pilosa Corp.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pilosa
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Default version check URL.
+const (
+	defaultVersionCheckURL = "https://diagnostics.pilosa.com/v0/version"
+)
+
+type versionResponse struct {
+	Version string `json:"version"`
+	Message string `json:"message"`
+}
+
+// DiagnosticsCollector represents a collector/sender of diagnostics data
+type DiagnosticsCollector struct {
+	mu          sync.Mutex
+	host        string
+	VersionURL  string
+	version     string
+	lastVersion string
+	startTime   int64
+	start       time.Time
+
+	metrics map[string]interface{}
+
+	client *http.Client
+
+	logOutput io.Writer
+
+	server *Server
+}
+
+// New returns a pointer to a new DiagnosticsCollector Client given an addr in the format "hostname:port".
+func NewDiagnosticsCollector(host string) *DiagnosticsCollector {
+
+	return &DiagnosticsCollector{
+		host:       host,
+		VersionURL: defaultVersionCheckURL,
+		startTime:  time.Now().Unix(),
+		start:      time.Now(),
+		client:     &http.Client{Timeout: 10 * time.Second},
+		metrics:    make(map[string]interface{}),
+		logOutput:  ioutil.Discard,
+	}
+}
+
+// SetVersion of locally running Pilosa Cluster to check against master.
+func (d *DiagnosticsCollector) SetVersion(v string) {
+	d.version = v
+	d.Set("Version", v)
+}
+
+// Flush sends the current metrics.
+func (d *DiagnosticsCollector) Flush() error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.metrics["Uptime"] = (time.Now().Unix() - d.startTime)
+	buf, err := d.encode()
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest("POST", d.host, bytes.NewReader(buf))
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := d.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// TODO verify response
```

Comment:
does this `TODO` need to be addressed now?

Reply:
I was wondering about that. I guess it depends on if we want to log that error or not. IMO, we shouldn't log it, as it does not affect pilosa's primary use or performance and presents no issue to the user. If you agree, I can just remove the TODO comment.

Reply:
yeah, i guess there's no reason to alert the user that diagnostics are failing to post. maybe just change the comment to indicate that we intentionally ignore this error.

Reply:
That's addressed now.